File: C++/문제_2512번.cpp
#include <iostream>
#include <algorithm>
using namespace std;

//문제 2512번
int main(){
    cin.tie(NULL);
    ios_base::sync_with_stdio(false);
    int n, i, num[10001], total, left = 0, right = 0 , mid,result = 0;
    cin >> n;
    for(i=0;i<n;i++){
        cin >> num[i];
        right = max(num[i], right);
    }
    cin >> total;
    while(left <= right){
        int sum = 0;
        mid = (left+right)/2;
        for(i=0;i<n;i++){
            if(mid >= num[i])
                sum+=num[i];
            else
                sum+= mid;
        }
        if(sum > total)
            right = mid - 1;
        else{
            left = mid + 1;
            result = mid;
        }
    }
    cout << result;
    return 0;
}

File: JavaScript/문제_2512번.js
let input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");
let n = +input[0];
input.shift();
let arr = input[0].split(' ').map(o => +o);
arr.sort((o1, o2) => {
  return o1 - o2;
});
let m = +input[1];
let result = +0, sum = +0;
let left = 0, right = arr[n-1];
while (left <= right) {
  let mid = parseInt((left + right) / 2);
  let innerSum = +0;
  arr.map((o) => {
    if (o >= mid)
      innerSum += mid;
    else
      innerSum += o;
  });
  if (sum < innerSum && m >= innerSum) {
    sum = innerSum;
    result = mid;
    left = mid + 1;
  }
  else
    right = mid - 1;
}
console.log(result);